File: hello-java/audio-stream-server/src/main/java/org/feuyeux/mmap/audio/server/handler/WebSocketMessage.java
package org.feuyeux.mmap.audio.server.handler;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * WebSocket message POJO for JSON serialization/deserialization.
 * Used for all control messages between client and server.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class WebSocketMessage {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    @JsonProperty("type")
    private String type;

    @JsonProperty("streamId")
    private String streamId;

    @JsonProperty("offset")
    private Long offset;

    @JsonProperty("length")
    private Integer length;

    @JsonProperty("message")
    private String message;

    // Default constructor for Jackson
    public WebSocketMessage() {
    }

    // All-args constructor
    public WebSocketMessage(String type, String streamId, Long offset, Integer length, String message) {
        this.type = type;
        this.streamId = streamId;
        this.offset = offset;
        this.length = length;
        this.message = message;
    }

    // Factory methods for common message types
    public static WebSocketMessage started(String streamId, String message) {
        return new WebSocketMessage(MessageType.STARTED.getValue(), streamId, null, null, message);
    }

    public static WebSocketMessage stopped(String streamId, String message) {
        return new WebSocketMessage(MessageType.STOPPED.getValue(), streamId, null, null, message);
    }

    public static WebSocketMessage connected(String streamId, String message) {
        return new WebSocketMessage(MessageType.CONNECTED.getValue(), streamId, null, null, message);
    }

    public static WebSocketMessage error(String message) {
        return new WebSocketMessage(MessageType.ERROR.getValue(), null, null, null, message);
    }

    /**
     * Get the message type as enum.
     *
     * @return MessageType enum, or null if type is not a valid value
     */
    public MessageType getMessageTypeEnum() {
        return MessageType.fromString(type);
    }

    /**
     * Parse JSON string to WebSocketMessage.
     */
    public static WebSocketMessage fromJson(String json) throws JsonProcessingException {
        return MAPPER.readValue(json, WebSocketMessage.class);
    }

    /**
     * Convert WebSocketMessage to JSON string.
     */
    public String toJson() throws JsonProcessingException {
        return MAPPER.writeValueAsString(this);
    }

    // Getters and Setters
    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getStreamId() {
        return streamId;
    }

    public void setStreamId(String streamId) {
        this.streamId = streamId;
    }

    public Long getOffset() {
        return offset;
    }

    public void setOffset(Long offset) {
        this.offset = offset;
    }

    public Integer getLength() {
        return length;
    }

    public void setLength(Integer length) {
        this.length = length;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    @Override
    public String toString() {
        return "WebSocketMessage{" +
                "type='" + type + '\'' +
                ", streamId='" + streamId + '\'' +
                ", offset=" + offset +
                ", length=" + length +
                ", message='" + message + '\'' +
                '}';
    }
}

File: hello-nodejs/src/server/handler/WebSocketMessage.js
/**
 * WebSocket message POJO for JSON serialization/deserialization.
 * Used for all control messages between client and server.
 */

const MessageType = require("./MessageType");

class WebSocketMessage {
  /**
   * Create a WebSocketMessage.
   *
   * @param {string} type - Message type
   * @param {string} streamId - Stream ID (optional)
   * @param {number} offset - Offset for GET requests (optional)
   * @param {number} length - Length for GET requests (optional)
   * @param {string} message - Additional message text (optional)
   */
  constructor(type, streamId = null, offset = null, length = null, message = null) {
    this.type = type;
    this.streamId = streamId;
    this.offset = offset;
    this.length = length;
    this.message = message;
  }

  /**
   * Factory method: Create a STARTED response message.
   *
   * @param {string} streamId - Stream ID
   * @param {string} message - Message text
   * @returns {WebSocketMessage}
   */
  static started(streamId, message = "Stream started successfully") {
    return new WebSocketMessage(MessageType.STARTED, streamId, null, null, message);
  }

  /**
   * Factory method: Create a STOPPED response message.
   *
   * @param {string} streamId - Stream ID
   * @param {string} message - Message text
   * @returns {WebSocketMessage}
   */
  static stopped(streamId, message = "Stream finalized successfully") {
    return new WebSocketMessage(MessageType.STOPPED, streamId, null, null, message);
  }

  /**
   * Factory method: Create a CONNECTED response message.
   *
   * @param {string} streamId - Connection ID
   * @param {string} message - Message text
   * @returns {WebSocketMessage}
   */
  static connected(streamId, message = "Connection established") {
    return new WebSocketMessage(MessageType.CONNECTED, streamId, null, null, message);
  }

  /**
   * Factory method: Create an ERROR response message.
   *
   * @param {string} message - Error message
   * @returns {WebSocketMessage}
   */
  static error(message) {
    return new WebSocketMessage(MessageType.ERROR, null, null, null, message);
  }

  /**
   * Get the message type as MessageType enum value.
   *
   * @returns {string|null} MessageType value, or null if type is not valid
   */
  getMessageType() {
    return MessageType.fromString(this.type);
  }

  /**
   * Parse JSON string to WebSocketMessage.
   *
   * @param {string} json - JSON string
   * @returns {WebSocketMessage}
   */
  static fromJson(json) {
    const data = JSON.parse(json);
    return new WebSocketMessage(
      data.type,
      data.streamId || null,
      data.offset || null,
      data.length || null,
      data.message || null
    );
  }

  /**
   * Convert WebSocketMessage to JSON string.
   * Excludes null values.
   *
   * @returns {string} JSON string
   */
  toJson() {
    const obj = {};
    if (this.type !== null) obj.type = this.type;
    if (this.streamId !== null) obj.streamId = this.streamId;
    if (this.offset !== null) obj.offset = this.offset;
    if (this.length !== null) obj.length = this.length;
    if (this.message !== null) obj.message = this.message;
    return JSON.stringify(obj);
  }

  /**
   * Convert to plain object (excluding null values).
   *
   * @returns {object}
   */
  toObject() {
    const obj = {};
    if (this.type !== null) obj.type = this.type;
    if (this.streamId !== null) obj.streamId = this.streamId;
    if (this.offset !== null) obj.offset = this.offset;
    if (this.length !== null) obj.length = this.length;
    if (this.message !== null) obj.message = this.message;
    return obj;
  }

  toString() {
    return `WebSocketMessage{type='${this.type}', streamId='${this.streamId}', offset=${this.offset}, length=${this.length}, message='${this.message}'}`;
  }
}

module.exports = WebSocketMessage;
